server/hwapi/data_models/repository.py

```diff
@@ -107,3 +111,253 @@ def get_or_create(
+                (
+                    models.Certificate.release_id == release.id
+                    if release is not None
+                    else true
+                ),
+            )
+        )
+        .distinct()
+        .all()
+    )
+
+
+def get_release_from_os(db: Session, os: OSValidator) -> models.Release:
+    """Return release object matching given OS data"""
+    release = (
+        db.query(models.Release)
+        .filter_by(release=os.version, codename=os.codename)
+        .first()
+    )
+    if release is None:
+        raise ValueError(
+            f"No matching release found for codename '{os.codename}', version '{os.version}'"
+        )
+    return release
+
+
+def clean_vendor_name(name):
+    """Remove "Inc"/"Inc." substring from vendor name and leading whitespaces"""
+    return name.replace("Inc.", "").replace("Inc", "").strip()
+
+
+def get_board_by_validator_data(
+    db: Session, board_validator: device_validators.BoardValidator
+) -> models.Device | None:
+    """Return device object (category==BOARD) matching given Board data"""
+    board = (
+        db.query(models.Device)
+        .join(models.Vendor)
+        .filter(
+            and_(
+                models.Vendor.name.in_(
+                    [
+                        board_validator.manufacturer,
+                        clean_vendor_name(board_validator.manufacturer),
+                    ]
+                ),
+                models.Device.name == board_validator.product_name,
+                models.Device.version == board_validator.version,
+                models.Device.category.in_(
+                    [DeviceCategory.BOARD.value, DeviceCategory.OTHER.value]
+                ),
+            )
+        )
+        .first()
+    )
+    return board
+
+
+def get_devices_by_machine_ids(db: Session, machine_ids: list[int]) -> Query:
+    """
+    Return query that contains Devices joined with Reports and Certificates for a
+    given list of machine IDs
+    """
+    return (
+        db.query(models.Device)
+        .join(
+            models.device_report_association,
+            models.Device.id == models.device_report_association.c.device_id,
+        )
+        .join(
+            models.Report,
+            models.device_report_association.c.report_id == models.Report.id,
+        )
+        .join(models.Certificate, models.Report.certificate_id == models.Certificate.id)
+        .filter(models.Certificate.machine_id.in_(machine_ids))
+    )
+
+
+def find_matching_processor(
+    db: Session,
+    devices_query: Query,
+    cpu_validator: device_validators.ProcessorValidator,
+) -> models.Device | None:
+    """
+    Find a CPU device across the given list of devices that contains
+    model name.
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.name.like(f"%{cpu_validator.version}%"),
+            models.Device.category == DeviceCategory.PROCESSOR,
+        ),
+    ).first()
```

Comment:
not too familiar with the device object or cpu data. But why is it that device name represents cpu version? Perhaps one needs to be renamed to the other.

Also couldn't like query produce false positives?